### PR TITLE
Add kexecbin pkg

### PR DESCRIPTION
**Pull request by zaolin**
* Add ARM64 and AMD64 statically compiled binary files.
* Add a wrapper around kexec-tools with dynamic runtime detection.

### CODE REVIEW

File: pkg/kexecbin/kexec_bin.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	// SecondDTBPath is the second dtp path to check
+	SecondDTBPath = "/proc/device-tree"
+)
+
+// KexecBin uses kexec-tools binary and runtime architecture detection
+// to execute abritary files.
+func KexecBin(kernelFilePath string, kernelCommandline string, initrdFilePath string, dtFilePath string) error {
+	var err error
+	var baseCmd string
+	switch runtime.GOARCH {
```
Comment (rjoleary) [inline pkg/kexecbin/kexec_bin.go]: You can do this at compile time:

- Have a global variable `var kexecPath string`
- In the file "kexecbin_amd64.go", have `func init { kexecPath = "kexec-amd64" }`
- In the file "kexecbin_arm64.go", have `func init { kexecPath = "kexec-arm64" }`

File: pkg/kexecbin/kexec_bin.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package kexecbin
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply pkg/kexecbin/kexec_bin.go]: Unless you intentionally want to make a build flag called "bin", you should name this file "kexecbin.go" (without the underscore).

File: pkg/kexecbin/kexec_bin.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	if initrdFilePath != "" {
+		loadCommands = append(loadCommands, "--initrd="+initrdFilePath)
+	}
+
+	if dtFilePath != "" {
+		loadCommands = append(loadCommands, "--dtb="+dtFilePath)
+	} else {
+		_, err := os.Stat(FirstDTBPath)
```
Comment (rjoleary) [inline pkg/kexecbin/kexec_bin.go]: Please turn this into a loop:

```
for _, dtFilePath := range dtFilePaths {
    _, err := os.Stat(dtFilePath)
    if err == nil {
          loadCommands = append(loadCommands, "--dtb="+dtFilePath)
          break
     }
}
```

File: pkg/kexecbin/kexec_bin.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"runtime"
+)
+
+var (
+	// FirstDTBPath is the first dtp path to check
+	FirstDTBPath = "/sys/firmware/fdt"
```
Comment (rjoleary) [inline pkg/kexecbin/kexec_bin.go]: Use a two element array to store both paths.

File: pkg/kexecbin/kexec_bin.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (zaolin) [reply pkg/kexecbin/kexec_bin.go]: Done

File: pkg/kexecbin/kexec_bin.go (hunk 1 above)
Comment (zaolin) [reply pkg/kexecbin/kexec_bin.go]: remove it. We just searching for kexec and during u-root initramfs generation users have to pass the right binary

File: pkg/kexecbin/kexec_bin.go (hunk 3 above)
Comment (zaolin) [reply pkg/kexecbin/kexec_bin.go]: Done

**Comment by zaolin** [comment]
Now it should be more the clean solution and https://github.com/linuxboot/bins delivers the binaries.

**Comment by hugelgupf** [comment]
> We'll need to get Ron's approval for this. His big thing is u-root has only 4-binaries. This adds two more binaries.

I don't think this adds binaries to u-root in principle. You can use this package if you want it, but you can also leave it. It's not tied to any commands right now.

**Comment by rminnich** [comment]
Let's talk about how to do this. I understand what you're trying to do here but there's way better ways to go. This is not it.

Also I'd like to understand why you need the utils binaries -- why can't we just fix our kexec?

**Comment by zaolin** [comment]
There is no other way for now and we want to provide support for it. Later on kexecbin will be removed by our own pure go implementation. This is just an additional package you don't have to use. It's pure Go and it works for us as long we can't provide our own implementation.

**Comment by insomniacslk** [comment]
I think it's good that we have a stop-gap until the Go kexec gets improved, but not convinced that the source code of the wrapper should live here. What about moving the code of this PR into github.com/u-root/bins, under cmds/kexecbin ? This wrapper and the underlying binaries are tightly coupled, and they should go together so it is obvious what people are opting in to

**Comment by rminnich** [comment]
I'd still like to know what's missing that you need but I guess we can let this in for now. But it's nasty.

**Comment by hugelgupf** [comment]
> I'd still like to know what's missing that you need but I guess we can let this in for now. But it's nasty.

Support for anything non-x86, or anything non-bzImage. Manual kexec, which I've been working on, works but is in the very very early stages.

**Pull request merged**